processing/src/main/java/io/druid/segment/data/GenericIndexed.java

```diff
@@ -147,31 +387,115 @@ public int indexOf(T value)
+      channel.write(ByteBuffer.wrap(new byte[]{VERSION_ONE, allowReverseLookup ? (byte) 0x1 : (byte) 0x0}));
+      channel.write(ByteBuffer.wrap(Ints.toByteArray(theBuffer.remaining() + 4)));
+      channel.write(ByteBuffer.wrap(Ints.toByteArray(size)));
+      channel.write(theBuffer.asReadOnlyBuffer());
+    } else {
+      throw new UnsupportedOperationException(
+          "GenericIndexed serialization for V2 is unsupported. Use GenericIndexedWriter instead.");
+    }
+  }
 
-  GenericIndexed(
-      ByteBuffer buffer,
-      ObjectStrategy<T> strategy,
-      boolean allowReverseLookup
-  )
+  /**
+   * Create a non-thread-safe Indexed, which may perform better than the underlying Indexed.
+   *
+   * @return a non-thread-safe Indexed
+   */
+  public GenericIndexed<T>.BufferIndexed singleThreaded()
   {
-    this.theBuffer = buffer;
-    this.strategy = strategy;
-    this.allowReverseLookup = allowReverseLookup;
+    final ByteBuffer copyHeaderBuffer = headerBuffer.asReadOnlyBuffer();
+
+    if (valueBuffers.size() == 1) {
+      final ByteBuffer copyBuffer = valueBuffers.get(0).asReadOnlyBuffer();
+      return new BufferIndexed()
+      {
+        @Override
+        public T get(final int index)
+        {
+          if (index < 0) {
+            throw new IAE("Index[%s] < 0", index);
+          }
+          if (index >= size) {
+            throw new IAE(String.format("Index[%s] >= size[%s]", index, size));
+          }
+
+          final int startOffset;
+          final int endOffset;
+
+          if (index == 0) {
+            copyHeaderBuffer.position(0);
+            startOffset = 4;
+            endOffset = copyHeaderBuffer.getInt();
+          } else {
+            copyHeaderBuffer.position((index - 1) * Ints.BYTES);
+            startOffset = copyHeaderBuffer.getInt() + 4;
+            endOffset = copyHeaderBuffer.getInt();
+          }
+          return _get(copyBuffer, startOffset, endOffset);
+        }
+      };
+    } else {
+      final List<ByteBuffer> copyValueBuffers = Lists.newArrayList(
+          Iterables.transform(
+              valueBuffers,
+              new Function<ByteBuffer, ByteBuffer>()
+              {
+                @Override
+                public ByteBuffer apply(final ByteBuffer buffer)
+                {
+                  return buffer.asReadOnlyBuffer();
+                }
+              }
+          )
+      );
+      return new BufferIndexed()
+      {
+        @Override
+        public T get(final int index)
+        {
+          int fileNum = index / numElementsPerValueFile;
+          final ByteBuffer copyBuffer = copyValueBuffers.get(fileNum);
+
+          if (index < 0) {
+            throw new IAE("Index[%s] < 0", index);
+          }
+          if (index >= size) {
+            throw new IAE(String.format("Index[%s] >= size[%s]", index, size));
+          }
+          final int startOffset;
+          final int endOffset;
+
+          if (index % numElementsPerValueFile == 0) {
```

Comment:
I kept both the blocks separately because version 1 does a check of if (index == 0) vs V2 check is if (index % numElementsPerValueFile == 0), we wanted to avoid % operation for version 1 indexed.